lib/request.ml

```diff
@@ -107,11 +107,14 @@ module Make(IO : S.IO) = struct
     | `Invalid reason as r -> return r
     | `Ok (meth, path, version) ->
       Header_IO.parse ic >>= fun headers ->
-      let uri = match (Header.get headers "host") with
+      let uri =
+        match Header.get headers "host" with
         | None -> Uri.of_string path
-        | Some h ->
-          (* XXX: Could be https:// as well... *)
-          Uri.of_string ("http://" ^ h ^ path) in
+        | Some host ->
+           let host_uri = Uri.of_string host in
+           Uri.of_string path |> fun uri ->
+           Uri.with_host uri (Uri.host host_uri)
```

Comment:
Isn't `host` just a bare hostname here? `Uri.of_string host` would then be a relative path...